### PR TITLE
Refactoring sunpy.net.attrs

### PR DESCRIPTION
**Pull request by Cadair**
The current state of `sunpy.net.attrs` is a bit of a mess as it has grown organically with the rest of net but never had a cleanup itself. The current state is that all the attrs are defined in `net/vso/attrs.py` and we import the ones which are generally applicable up into the `sunpy.net.attrs` namespace.

This PR makes the following changes:

### vso.attrs

All the classes defined in `vso.attrs` which are used outside of the VSO client have been moved into `net/_attrs.py`.

### `AttrWalker`

The `AttrWalker` class is designed to parse a query (collection of `Attr`s) and return a type specific to the client. This PR adds some limited inline documentation (I am planning a followup PR with narrative docs), as well as a basic set of API tests.

The `AttrWalker` was also the last place in sunpy still using `sunpy/util/multimethod` which I have refactored to use `functools.singledispatch` and a modified version thereof which dispatches on the second argument to the function and not the first.

I have now deprecated the use of `MultiMethod`.

### Generic cleanup

* `_Range` is no longer a mixin class but a subclass of `Attr` and should be used accordingly.
* I have endeavoured to cleanup the various `attrs.` namespaces to be free from all other variables so that your tab completion experience is as nice as possible.

## To Do:

Currently there is no deprecation of the classes that used to be in the `vso.attrs` namespace. Before I deprecate these we need to decide if the goal is to have the client specific attrs be only classes used exclusively by those clients (as it is now in this PR) or all the ones applicable to those clients (i.e. also including the generic ones) which would mean this PR isn't an API change for vso at all.

### CODE REVIEW

File: sunpy/net/attr.py [hunk 1]
```diff
@@ -429,9 +429,12 @@ class AttrWalker:
       type and return a different one.
 
     """
+    @staticmethod
+    def _unknown_type(*args, **kwargs):
+        raise TypeError(f"{args[1]} or any of its parents have not been registered with the AttrWalker")
     def __init__(self):
```
Comment (Cadair) [inline sunpy/net/attr.py]:
```suggestion

    def __init__(self):
```

File: sunpy/net/vso/attrs.py [hunk 2]
```diff
@@ -536,34 +282,37 @@ def _create(wlk, root, api):
 # AttrAnd and AttrOr obviously are - in the HEK module). If we defined the
 # filter method as a member of the attribute classes, we could only filter
 # one type of data (that is, VSO data).
-filter_results = MultiMethod(lambda *a, **kw: (a[0], ))
+
+@_singledispatch
+def _filter_results(*args, **kwargs):
```
Comment (Cadair) [inline sunpy/net/vso/attrs.py]: Well we don't have any tests for this, so hopefully it's not broken :see_no_evil:

File: sunpy/net/hek2vso/hek2vso.py [hunk 3]
```diff
@@ -92,12 +93,12 @@ def vso_attribute_parse(phrase):
     [<Time(<Time object: scale='utc' format='isot' value=2011-08-09T07:22:38.000>, <Time object: scale='utc' format='isot' value=2011-08-09T08:32:02.000>, None)>, <Source('SDO')>, <Instrument('AIA')>, <Wavelength(210.99999999999997, 210.99999999999997, 'Angstrom')>]
     """
     try:
-        query = [vso.attrs.Time(phrase['event_starttime'],
+        query = [a.Time(phrase['event_starttime'],
                                 phrase['event_endtime']),
-                 vso.attrs.Source(phrase['obs_observatory']),
-                 vso.attrs.Instrument(phrase['obs_instrument'])]
+                 a.vso.Source(phrase['obs_observatory']),
```
Comment (nabobalis) [inline sunpy/net/hek2vso/hek2vso.py]: Source is still vso specific?

File: sunpy/util/functools.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import functools
+
+
+__all__ = ['seconddispatch']
+
+
+def seconddispatch(func):
+    """
+    A variant of `functools.singledispatch` which dispatches on type of the second argument.
+
```
Comment (nabobalis) [inline sunpy/util/functools.py]:
```suggestion
```

**Comment by Cadair** [comment]
Discussed at the community meeting, should deprecate names in `vso.attrs`.

File: sunpy/net/vso/attrs.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/net/vso/attrs.py]: You add one?

File: sunpy/net/attr.py [hunk 5]
```diff
@@ -261,10 +261,49 @@ def __repr__(self):
...
+    An attribute that represents a range of a value.
+
+    This type of attribute would be applicable for types like Wavelength or Time.
+    The range is inclusive of both the min and ma
+
+    Parameters
+    ----------
+    min_ : `object`
+        The lower bound of the range.
+
```
Comment (nabobalis) [inline sunpy/net/attr.py]:
```suggestion
```

File: sunpy/net/_attrs.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    pass
+
+
+class Resolution(SimpleAttr):
+    """
+    Resolution level of the data.
+
+    Parameters
+    ----------
+    value : float or str
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
    value : `float` or `str`
```

File: sunpy/net/_attrs.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        self.value = value.to_value(u.s)
+
+
+class Detector(SimpleAttr):
+    """
+    The detector from which the data comes from.
+
+    Parameters
+    ----------
+    value : str
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
    value : `str`
```

File: sunpy/net/_attrs.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+
+class Level(SimpleAttr):
+    """
+    Specifies the data processing level to search for.  The data processing
+    level is specified by the instrument PI.  May not work with all archives.
+
+    Parameters
+    ----------
+    value : float or str
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
    value : `float` or `str`
```

File: sunpy/net/_attrs.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+                                                            self.unit)
+
+
+class Instrument(SimpleAttr):
+    """
+    Specifies the Instrument name for the search.
+
+    Parameters
+    ----------
+    value : str
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
    value : `str`
```

File: sunpy/net/_attrs.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+    Notes
+    -----
+    More information about each instrument supported by the VSO may be found
+    within the VSO Registry. For a list of instruments see
+    https://sdac.virtualsolar.org/cgi/show_details?keyword=INSTRUMENT.
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]: Turn into a `.. note::` section?

File: sunpy/net/_attrs.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+        wavemax : `~astropy.units.Quantity`
+            The upper bound of the range, if not specified it will default to
+            the lower bound.
+
+        Notes
+        -----
+        The VSO understands the 'wavelength' in one of three units, Angstroms,
+        kHz or keV. Therefore any unit which is directly convertible to these
+        units is valid input.
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]: Turn into a `.. note::`?

File: sunpy/net/_attrs.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+class Wavelength(Range):
+    def __init__(self, wavemin, wavemax=None):
+        """
+        Specifies the wavelength or spectral energy range of the detector.
+
+        Parameters
+        ----------
+        wavemin : `~astropy.units.Quantity`
+            The lower bounds of the range.
+
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
```

File: sunpy/net/_attrs.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+    end : SunPy Time String
+        The end time of the range.
+
+    near : SunPy Time String
+        Return a singular record closest in time to this value as possible,
+        inside the start and end window. Note: not all providers support this
+        functionality.
+
+    """
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]: Update this to use the new time doc add?

File: sunpy/net/_attrs.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+    Parameters
+    ----------
+    start : SunPy time string or `~sunpy.time.TimeRange`.
+        The start time in a format parseable by `~sunpy.time.parse_time` or
+        a `sunpy.time.TimeRange` object.
+
+    end : SunPy Time String
+        The end time of the range.
+
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
```

File: sunpy/net/_attrs.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+class Time(Range):
+    """
+    Specify the time range of the query.
+
+    Parameters
+    ----------
+    start : SunPy time string or `~sunpy.time.TimeRange`.
+        The start time in a format parseable by `~sunpy.time.parse_time` or
+        a `sunpy.time.TimeRange` object.
+
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
```

File: sunpy/net/hek2vso/hek2vso.py (hunk 3 above)
Comment (Cadair) [reply sunpy/net/hek2vso/hek2vso.py]: for now.

**Comment by Cadair** [comment]
> I have a silly question - why is it necessary to put from sunpy.net import attrs as a within the _can_handle_query() for the dataretriever clients (goes and norh)? How will this prevent circular imports?

The issue is that the clients are imported by the `net/__init__.py` file, which also imports the attrs. So when the `net/dataretriever/goes/` files are imported the `sunpy.net.attrs` namespace doesn't exist yet, because we are still importing `sunpy.net`.

Or to put it another way `sunpy.net` imports `sunpy.net.dataretriever.goes` which imports `sunpy.net.attrs` which breaks because `sunpy.net` doesn't exist.

 By putting the imports in the `_can_handle_query` method importing `sunpy.net.dataretriever` doesn't import `sunpy.net.attrs` that import happens at runtime.

File: sunpy/net/vso/attrs.py (hunk 2 above)
Comment (Cadair) [reply sunpy/net/vso/attrs.py]: I would rather fold it into https://github.com/sunpy/sunpy/issues/3733 really.

**Comment by Cadair** [comment]
Tests are fine other than the HEK having got upset.

File: sunpy/net/vso/attrs.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/net/vso/attrs.py]: Fine by me.

File: sunpy/net/vso/attrs.py [hunk 16]
```diff
@@ -619,3 +370,24 @@ def _(attr, results):
...
+
+for _name in _deprecated_names:
+    # Dynamically construct a class which inherits the class with the
+    # deprecation warning in the __init__ first and the class it's deprecating
+    # second.
+    cls = type(_name, (_DeprecatedAttr, getattr(_attrs, _name)), {})
+    # Add the new class to the modules namespace
+    setattr(sys.modules[__name__], _name, cls)
+
+__all__ += _deprecated_names
```
Comment (nabobalis) [inline sunpy/net/vso/attrs.py]: 10/10

**Pull request merged**